geo/src/algorithm/concavehull.rs

```diff
@@ -0,0 +1,418 @@
+
+pub trait ConcaveHull<T> {
+    /// Returns the concave hull of a Polygon.
+    ///
+    /// This implementation is inspired by https://github.com/mapbox/concaveman
+    /// and also uses ideas from the following paper:
+    /// www.iis.sinica.edu.tw/page/jise/2012/201205_10.pdf
+    ///
+    /// # Example
+    /// ```
+    /// use geo::{line_string, polygon};
+    /// use geo::algorithm::concavehull::ConcaveHull;
+    ///
+    /// // a square shape
+    /// let poly = polygon![
+    ///     (x: 0.0, y: 0.0),
+    ///     (x: 4.0, y: 0.0),
+    ///     (x: 4.0, y: 4.0),
+    ///     (x: 0.0, y: 4.0),
+    /// ];
+    ///
+    /// // The correct concave hull coordinates
+    /// let correct_hull = line_string![
+    ///     (x: 4.0, y: 0.0),
+    ///     (x: 4.0, y: 4.0),
+    ///     (x: 0.0, y: 4.0),
+    ///     (x: 0.0, y: 0.0),
+    ///     (x: 4.0, y: 0.0),
+    /// ];
+    ///
+    /// let res = poly.concave_hull(2.0);
+    /// assert_eq!(res.exterior(), &correct_hull);
+    /// ```
+    fn concave_hull(&self, concavity: T) -> Polygon<T>
+    where
+        T: Float + RTreeNum;
+}
+
+impl<T> ConcaveHull<T> for Polygon<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let points: Vec<_> = self.exterior().points_iter().collect();
+        Polygon::new(LineString::from(concave_hull(&points, concavity)), vec![])
+    }
+}
+
+impl<T> ConcaveHull<T> for MultiPolygon<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let aggregated: Vec<Point<T>> = self
+            .0
+            .iter()
+            .flat_map(|elem| elem.exterior().0.iter().map(|c| Point(*c)))
+            .collect();
+        Polygon::new(
+            LineString::from(concave_hull(&aggregated, concavity)),
+            vec![],
+        )
+    }
+}
+
+impl<T> ConcaveHull<T> for LineString<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        Polygon::new(
+            LineString::from(concave_hull(&self.clone().into_points(), concavity)),
+            vec![],
+        )
+    }
+}
+
+impl<T> ConcaveHull<T> for MultiLineString<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let aggregated: Vec<Point<T>> = self
+            .0
+            .iter()
+            .flat_map(|elem| elem.clone().into_points())
+            .collect();
+        Polygon::new(
+            LineString::from(concave_hull(&aggregated, concavity)),
+            vec![],
+        )
+    }
+}
+
+impl<T> ConcaveHull<T> for MultiPoint<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        Polygon::new(
+            LineString::from(concave_hull(&self.0.clone(), concavity)),
+            vec![],
+        )
+    }
+}
+
+fn find_point_closest_to_line<T>(
+    interior_points_tree: &RTree<Point<T>>,
+    line: Line<T>,
+    max_dist: &T,
```

Comment:
```suggestion
    max_dist: T,
```

Reply:
I've never used GH code "suggestion" feature... let's see if it works, I tried to suggest the way that I would write this bit. T is Float, so that means it's also `Copy`. I'd probably lean into that here and not borrow T at all.